applications/RomApplication/custom_strategies/lspg_rom_builder_and_solver.h

```diff
@@ -0,0 +1,587 @@
+        BuildROM(pScheme, rModelPart, Arom, brom);
+
+        // Obtain the assembled residuals vector (To build a basis for Petrov-Galerkin)
+        if (mTrainPetrovGalerkinFlag){
+            Kratos::Vector r_residual;
+            GetAssembledResiduals(pScheme, rModelPart, r_residual);
+        }
+
+        SolveROM(rModelPart, Arom, brom, Dx);
+
+
+        KRATOS_CATCH("")
+    }
+
+    Parameters GetDefaultParameters() const override
+    {
+        Parameters default_parameters = Parameters(R"(
+        {
+            "name" : "lspg_rom_builder_and_solver",
+            "nodal_unknowns" : [],
+            "number_of_rom_dofs" : 10,
+            "train_petrov_galerkin" : false
+        })");
+        default_parameters.AddMissingParameters(BaseType::GetDefaultParameters());
+
+        return default_parameters;
+    }
+
+    static std::string Name() 
+    {
+        return "lspg_rom_builder_and_solver";
+    }
+
+    ///@}
+    ///@name Access
+    ///@{
+
+
+    ///@}
+    ///@name Inquiry
+    ///@{
+
+
+    ///@}
+    ///@name Input and output
+    ///@{
+
+    /// Turn back information as a string.
+    virtual std::string Info() const override
+    {
+        return "LSPGROMBuilderAndSolver";
+    }
+
+    /// Print information about this object.
+    virtual void PrintInfo(std::ostream &rOStream) const override
+    {
+        rOStream << Info();
+    }
+
+    /// Print object's data.GetNumberOfROMModes()
+    virtual void PrintData(std::ostream &rOStream) const override
+    {
+        rOStream << Info();
+    }
+
+    ///@}
+    ///@name Friends
+    ///@{
+
+
+    ///@}
+protected:
+
+    SizeType mNodalDofs;
+    bool mTrainPetrovGalerkinFlag = false;
+
+    ///@}
+    ///@name Protected static member variables
+    ///@{
+
+
+    ///@}
+    ///@name Protected member variables
+    ///@{
+
+    ///@}
+    ///@name Protected operators
+    ///@{
+
+
+    ///@}
+    ///@name Protected operations
+    ///@{
+
+    void AssignSettings(const Parameters ThisParameters) override
+    {
+        BaseType::AssignSettings(ThisParameters);
+
+        // // Set member variables
+        mTrainPetrovGalerkinFlag = ThisParameters["train_petrov_galerkin"].GetBool();
+    }
+
+    /**
+    * Thread Local Storage containing dynamically allocated structures to avoid reallocating each iteration.
+    */
+    struct AssemblyTLS
+    {
+        Matrix phiE = {};                // Elemental Phi
+        LocalSystemMatrixType lhs = {};  // Elemental LHS
+        EquationIdVectorType eq_id = {}; // Elemental equation ID vector
+        DofsVectorType dofs = {};        // Elemental dof vector
+        RomSystemMatrixType romA;        // reduced LHS
+        RomSystemVectorType romB;        // reduced RHS
+    };
+
+    /**
+     * Resizes a Matrix if it's not the right size
+     */
+    template<typename TMatrix>
+    static void ResizeIfNeeded(TMatrix& mat, const SizeType rows, const SizeType cols)

```

Comment:
```suggestion
    static void ResizeIfNeeded(TMatrix& rMat, const SizeType Rows, const SizeType Cols)

```
Style guide convention (CamelCase for method arguments).